padertorch/contrib/examples/speaker_classification/supervised/data.py

```diff
@@ -0,0 +1,118 @@
+import numpy as np
+import torch
+
+from lazy_dataset.database import JsonDatabase
+
+import paderbox as pb
+import padertorch as pt
+from padertorch.ops.sequence.pack_module import pad_sequence
+from padertorch.contrib.je.data.transforms import LabelEncoder
+
+
+def get_datasets(
+    storage_dir, database_json, dataset, batch_size=16, prefetch=True
+):
+    db = JsonDatabase(database_json)
+    ds = db.get_dataset(dataset)
+
+    def prepare_example(example):
+        example['audio_path'] = example['audio_path']['observation']
+        example['speaker_id'] = example['speaker_id'].split('-')[0]
+        return example
+
+    ds = ds.map(prepare_example)
+
+    speaker_encoder = LabelEncoder(
+        label_key='speaker_id', storage_dir=storage_dir, to_array=True
+    )
+    speaker_encoder.initialize_labels(dataset=ds, verbose=True)
+    ds = ds.map(speaker_encoder)
+
+    train_set, validate_set, test_set = train_test_split(ds)
+    training_data = prepare_dataset(train_set, batch_size, training=True)
+    validation_data = prepare_dataset(validate_set, batch_size, training=False)
+    test_data = prepare_dataset(
+        test_set, batch_size, training=False, prefetch=prefetch
+    )
+    return training_data, validation_data, test_data
+
+
+def train_test_split(dataset, dev_split=0.1,  test_split=0.1, seed=0):
+    r = np.random.RandomState(seed)
+
+    try:
+        num_examples = len(dataset)
+    except TypeError:
+        raise RuntimeError('dataset must be indexable!')
+
+    indices = np.arange(num_examples)
+
+    dev_size = int(num_examples * dev_split)
+    test_size = int(num_examples * test_split)
+
+    test_candidates = r.choice(indices, size=test_size, replace=False)
+    indices = np.delete(indices, test_candidates)
+    dev_candidates = r.choice(indices, size=dev_size, replace=False)
+    train_candidates = np.delete(indices, dev_candidates)
+
+    return (
+        dataset[train_candidates], dataset[dev_candidates],
+        dataset[test_candidates]
+    )
+
+
+def _prepare_features(example):
+    audio_data = pb.io.load_audio(
+        example['audio_path'], expected_sample_rate=16000
+    )
+    stft_transform = pb.transform.STFT(
+        shift=160, window_length=400, size=512, fading=None, pad=False
+    )
+    stft = stft_transform(audio_data)
+    mel_transform = pb.transform.module_fbank.MelTransform(
+        sample_rate=16000, fft_length=512, n_mels=64, fmin=50
+    )
+    example['mel_transform'] = mel_transform(np.abs(stft) ** 2)
+    return example
+
+
+def _collate_example(example):
+    example = pt.data.utils.collate_fn(example)
+    example['features'] = pad_sequence(
+        example['features'], batch_first=True
+    ).transpose(-1, -2)
+    example['speaker_id'] = np.stack(example['speaker_id'])
+    return example
+
+
+def prepare_dataset(dataset, batch_size=16, training=False, prefetch=True):
+    dataset = dataset.map(_prepare_features)
```

Comment:
Is it not easier to read the code, when you do only one mapping to the dataset?

Reply:
https://github.com/fgnt/padertorch/pull/57/commits/3654cbd02a18b9be271f20ea1898ccd5c4cc3f81